saleor/order/actions.py

```diff
@@ -1361,37 +1370,42 @@ def _process_refund(
     refund_amount = _calculate_refund_amount(
         order_lines_to_refund, fulfillment_lines_to_refund, lines_to_refund
     )
+
+    # Happens when neither payments_to_refund nor amount_to_refund were specified.
     if amount is None:
         amount = refund_amount
         # we take into consideration the shipping costs only when amount is not
         # provided.
         if refund_shipping_costs:
             amount += order.shipping_price_gross_amount
+        # At this point there can be only one payment.
+        payments[0]["amount"] = min(payments[0]["payment"].captured_amount, amount)
+
     if amount:
-        amount = min(payment.captured_amount, amount)
-        try:
-            gateway.refund(
-                payment, manager, amount=amount, channel_slug=order.channel.slug
-            )
-        except PaymentError:
-            raise ValidationError(
-                "The refund operation is not available yet.",
-                code=OrderErrorCode.CANNOT_REFUND.value,
-            )
-        transaction.on_commit(
-            lambda: events.payment_refunded_event(
+        for item in payments:
+            try_refund(
                 order=order,
                 user=user,
                 app=app,
-                amount=amount,  # type: ignore
-                payment=payment,
+                payment=item["payment"],
+                manager=manager,
+                channel_slug=order.channel.slug,
```

Comment:
`order.channel` is a relationship, not sure if it wouldn't run a database query in each iteration? we could assign the channel slug to a variable before this for-loop and use the variable here.

Reply:
Django uses a cached value for later calls.
From [the docs](https://docs.djangoproject.com/en/3.2/topics/db/queries/#one-to-many-relationships):

> Forward access to one-to-many relationships is cached the first time the related object is accessed. Subsequent accesses to the foreign key on the same object instance are cached.